### PR TITLE
Add JSDoc annotations to ensure enums type check even without `.d.ts` file

### PR DESCRIPTION
**Pull request by RunDevelopment**
This is a partial fix for #3864.

I added additional JSDoc annotations to ensure that enums type check even without explicit type definition (`.d.ts` file). 

C-style enums get an `@enum {1|2|3|4}` annotation which (1) marks the frozen enum object as an enum and (2) ensures that type annotations properly resolve. A comment in #3864 suggested using `@enum {number}`, but I defined the enum as the union of all possible numeric values instead to make the type more specific.

String enums get `@typedef {"a" | "b" | "c"} StringEnum` which is exactly the same as TS `type StringEnum = "a" | "b" | "c"`. This means that string enums have exact type definitions with and without a `.d.ts` file. I also added `@type {StringEnum[]}` to arrays of string enum variants to ensure that the internal uses of those arrays don't cause type errors.

With those 2 changes, all of `enum.js` type checks without its `.d.ts` file.

One remaining problem is that I don't know how to test for the lack of type errors. While making this PR, I just added a `// @ts-check` at the top of `enum.js` so I could see type errors in my IDE. Any ideas how I could add tests for this?

### CODE REVIEW

**Comment by daxpedda** [comment]
@RunDevelopment if you rebase this PR, I'm happy to give it a review.

> One remaining problem is that I don't know how to test for the lack of type errors. While making this PR, I just added a `// @ts-check` at the top of `enum.js` so I could see type errors in my IDE. Any ideas how I could add tests for this?

Unfortunately not.
I have almost no experience with JS/TS linting, but ideally all our output testing should be covered by them.

**Comment by RunDevelopment** [comment]
Rebased. Thank you for reviewing!

File: crates/cli/tests/reference/enums.js [hunk 1]
```diff
@@ -78,9 +79,16 @@ Yellow:1,"1":"Yellow",
...
  */
 Red:2,"2":"Red", });
-
+/**
+ * @enum {0 | 1 | 42 | 43}
+ */
 export const ImplicitDiscriminant = Object.freeze({ A:0,"0":"A",B:1,"1":"B",C:42,"42":"C",D:43,"43":"D", });
 
+/**
+ * The name of a color.
```
Comment (daxpedda) [inline crates/cli/tests/reference/enums.js:88]: Considering this isn't exported, I believe doc comments shouldn't be exposed as well.
WDYT?

File: crates/cli/tests/reference/enums.js [hunk 2]
```diff
@@ -78,9 +79,16 @@ Yellow:1,"1":"Yellow",
...
+ * @typedef {"green" | "yellow" | "red"} ColorName
```
Comment (daxpedda) [inline crates/cli/tests/reference/enums.js:89]: Does this get registered even though the user only gets to use and see `get_name()`?

File: crates/cli/tests/reference/enums.js [hunk 3]
```diff
@@ -78,9 +79,16 @@ Yellow:1,"1":"Yellow",
...
+ * @type {ColorName[]}
```
Comment (daxpedda) [inline crates/cli/tests/reference/enums.js:90]: > I also added `@type {StringEnum[]}` to arrays of string enum variants to ensure that the internal uses of those arrays don't cause type errors.

This belong in #4189 then.

File: crates/cli/tests/reference/enums.js (hunk 1 above)
Comment (RunDevelopment) [reply crates/cli/tests/reference/enums.js:88]: I think that's reasonable. I'll remove them.

**Review by daxpedda — CHANGES_REQUESTED** [review]
LGTM!
Missing a changelog entry.

**Comment by RunDevelopment** [comment]
Done, although I'm not sure whether the explanation I picked for this change is good...

**Review by daxpedda — APPROVED** [review]
Would appreciate if you could do the rebase!

**Comment by RunDevelopment** [comment]
Sure thing!

**Pull request merged**